src/rust/inetstack/protocols/udp/queue.rs

```diff
@@ -121,65 +181,117 @@ impl<T> SharedQueue<T> {
 }
 
 /// Getters and setters for per UDP queue metadata.
-impl UdpQueue {
-    pub fn new() -> Self {
+impl<const N: usize> UdpQueue<N> {
+    pub fn new(rt: Rc<dyn NetworkRuntime<N>>, recv_queue: SharedQueue<N>) -> Self {
         Self {
-            addr: None,
-            recv_queue: None,
+            local: None,
+            rt,
+            recv_queue,
         }
     }
+}
 
-    /// Check whether the queue/socket is bound to an address.
-    pub fn is_bound(&self) -> bool {
-        self.addr != None
+impl<const N: usize> SharedUdpQueue<N> {
+    pub fn new(rt: Rc<dyn NetworkRuntime<N>>, recv_queue: SharedQueue<N>) -> Self {
+        Self(SharedObject::new(UdpQueue::new(rt, recv_queue)))
     }
 
-    /// Get the address assigned to this socket.
-    pub fn get_addr(&self) -> Result<SocketAddrV4, Fail> {
-        match self.addr {
-            Some(addr) => Ok(addr.clone()),
-            None => Err(Fail::new(libc::ENOTCONN, "port not bound")),
-        }
+    pub fn bind(&mut self, local: SocketAddrV4) -> Result<(), Fail> {
+        self.local = Some(local);
+        Ok(())
+    }
+
+    pub fn close(&self) -> Result<(), Fail> {
+        Ok(())
     }
 
-    /// Get the recv queue associated with this socket.
-    pub fn get_recv_queue(&self) -> SharedQueue<SharedQueueSlot<DemiBuffer>> {
-        match &self.recv_queue {
-            Some(recv) => recv.clone(),
-            None => panic!("No allocated receive queue!"),
+    pub fn pushto(
+        &self,
+        local_ipv4_addr: Ipv4Addr,
+        remote: &SocketAddrV4,
+        local_link_addr: MacAddress,
+        remote_link_addr: MacAddress,
+        buf: DemiBuffer,
+        offload_checksum: bool,
+    ) -> Result<(), Fail> {
+        // Fast path: try to send the datagram immediately.
+        let udp_header: UdpHeader = UdpHeader::new(self.local.expect("socket should be bound").port(), remote.port());
+        debug!("UDP send {:?}", udp_header);
+        let datagram = UdpDatagram::new(
+            Ethernet2Header::new(remote_link_addr, local_link_addr, EtherType2::Ipv4),
+            Ipv4Header::new(local_ipv4_addr, remote.ip().clone(), IpProtocol::UDP),
+            udp_header,
+            buf,
+            offload_checksum,
+        );
+        Ok(self.rt.transmit(Box::new(datagram)))
+    }
+
+    pub async fn pop(&mut self, size: Option<usize>, yielder: Yielder) -> Result<(SocketAddrV4, DemiBuffer), Fail> {
+        const MAX_POP_SIZE: usize = 9000;
```

Comment:
What is the rationale for this magic constant? Shouldn't it be in the libos layer?